File: CR.Python-GRPC-Adenine/elastos_adenine/sidechain_eth.py
import json
import grpc
import jwt
import datetime
from solidity_parser import parser

from .stubs import sidechain_eth_pb2, sidechain_eth_pb2_grpc
from elastos_adenine.settings import REQUEST_TIMEOUT, TOKEN_EXPIRATION


class SidechainEth:

    def __init__(self, host, port, production):
        if not production:
            self._channel = grpc.insecure_channel('{}:{}'.format(host, port))
        else:
            credentials = grpc.ssl_channel_credentials()
            self._channel = grpc.secure_channel('{}:{}'.format(host, port), credentials)

        self.stub = sidechain_eth_pb2_grpc.SidechainEthStub(self._channel)

    def close(self):
        self._channel.close()

    def deploy_eth_contract(self, api_key, did, network, eth_account_address, eth_private_key, eth_gas, filename):
        with open(filename, 'r') as myfile:
            contract_source = myfile.read()
        contract_metadata = parser.parse_file(filename)
        contract_name = contract_metadata['children'][1]['name']
        
        req_data = {
            'eth_account_address': eth_account_address,
            'eth_private_key': eth_private_key,
            'eth_gas': eth_gas,
            'contract_source': contract_source,
            'contract_name': contract_name,
        }

        jwt_info = {
            'network': network,
            'request_input': req_data
        }

        jwt_token = jwt.encode({
            'jwt_info': jwt_info,
            'exp': datetime.datetime.utcnow() + datetime.timedelta(hours=TOKEN_EXPIRATION)
        }, api_key, algorithm='HS256')

        response = self.stub.DeployEthContract(sidechain_eth_pb2.Request(input=jwt_token), timeout=REQUEST_TIMEOUT, metadata=[('did', did)])
        
        if response.status:
            output = jwt.decode(response.output, key=api_key, algorithms=['HS256']).get('jwt_info')
            result = {
                'output': json.dumps(output),
                'status_message': response.status_message,
                'status': response.status
            }
            return result
        else:
            result = {
                'output': '',
                'status_message': response.status_message,
                'status': response.status
            }
            return result

    def watch_eth_contract(self, api_key, did, network, contract_address, contract_name, contract_code_hash):
        req_data = {
            'contract_address': contract_address,
            'contract_name': contract_name,
            'contract_code_hash': contract_code_hash,
        }

        jwt_info = {
            'network': network,
            'request_input': req_data
        }

        jwt_token = jwt.encode({
            'jwt_info': jwt_info,
            'exp': datetime.datetime.utcnow() + datetime.timedelta(hours=TOKEN_EXPIRATION)
        }, api_key, algorithm='HS256')

        response = self.stub.WatchEthContract(sidechain_eth_pb2.Request(input=jwt_token), timeout=REQUEST_TIMEOUT, metadata=[('did', did)])
        
        if response.status:
            output = jwt.decode(response.output, key=api_key, algorithms=['HS256']).get('jwt_info')
            result = {
                'output': json.dumps(output),
                'status_message': response.status_message,
                'status': response.status
            }
            return result
        else:
            result = {
                'output': '',
                'status_message': response.status_message,
                'status': response.status
            }
            return result

File: CR.Go-GRPC-Adenine/elastosadenine/sidechain_eth.go
package elastosadenine

import (
	"context"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"log"
	"time"

	"github.com/antlr/antlr4/runtime/Go/antlr"
	"github.com/cyber-republic/go-grpc-adenine/v2/elastosadenine/stubs/sidechain_eth"
	"github.com/dgrijalva/jwt-go"
	"github.com/susmit/Solidity-Go-Parser/parser"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials"
	"google.golang.org/grpc/metadata"
)

type SidechainEth struct {
	Connection *grpc.ClientConn
}

type JWTInfoSidechainEthDeploy struct {
	Network        string `json:"network"`
	Address        string `json:"eth_account_address"`
	PrivateKey     string `json:"eth_private_key"`
	Gas            int    `json:"eth_gas"`
	ContractSource string `json:"contract_source"`
	ContractName   string `json:"contract_name"`
}

type JWTInfoSidechainEthWatch struct {
	Network          string `json:"network"`
	ContractAddress  string `json:"contract_address"`
	ContractName     string `json:"contract_name"`
	ContractCodeHash string `json:"contract_code_hash"`
}

type SolidityListener struct {
	*parser.BaseSolidityListener
	ContractName string
}

func (s *SolidityListener) EnterContractDefinition(ctx *parser.ContractDefinitionContext) {
	s.ContractName = fmt.Sprint(ctx.GetChildren()[1].GetChildren()[0])
}

func NewSidechainEth(host string, port int, production bool) *SidechainEth {
	address := fmt.Sprintf("%s:%d", host, port)
	var opts []grpc.DialOption
	if production == false {
		opts = []grpc.DialOption{
			grpc.WithInsecure(),
			grpc.WithMaxMsgSize(grpcMaxMsgSize),
		}
	} else {
		config := &tls.Config{
			InsecureSkipVerify: true,
		}
		opts = []grpc.DialOption{
			grpc.WithTransportCredentials(credentials.NewTLS(config)),
			grpc.WithMaxMsgSize(grpcMaxMsgSize),
		}
	}
	conn, err := grpc.Dial(address, opts...)
	if err != nil {
		log.Fatalf("Failed to connect to gRPC server: %v", err)
	}
	return &SidechainEth{Connection: conn}
}

func (e *SidechainEth) Close() {
	e.Connection.Close()
}

func (e *SidechainEth) DeployEthContract(apiKey, did, network, address, privateKey string, gas int, fileName string) Response {
	var outputData string
	client := sidechain_eth.NewSidechainEthClient(e.Connection)

	// Parse the solidity smart contract
	contractSource, _ := antlr.NewFileStream(fileName)
	lexer := parser.NewSolidityLexer(contractSource)
	stream := antlr.NewCommonTokenStream(lexer, 0)
	p := parser.NewSolidityParser(stream)
	solidityListener := SolidityListener{}
	antlr.ParseTreeWalkerDefault.Walk(&solidityListener, p.SourceUnit())

	jwtInfo, _ := json.Marshal(JWTInfoSidechainEthDeploy{
		Network:        network,
		Address:        address,
		PrivateKey:     privateKey,
		Gas:            gas,
		ContractSource: fmt.Sprint(contractSource),
		ContractName:   solidityListener.ContractName,
	})

	claims := JWTClaim{
		JwtInfo: string(jwtInfo),
		StandardClaims: jwt.StandardClaims{
			ExpiresAt: time.Now().UTC().Add(tokenExpiration).Unix(),
		},
	}

	jwtToken := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	jwtTokenString, err := jwtToken.SignedString([]byte(apiKey))
	if err != nil {
		log.Fatalf("Failed to execute 'DeployEthContract' method: %v", err)
	}
	md := metadata.Pairs("did", did)
	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
	defer cancel()
	ctx = metadata.NewOutgoingContext(ctx, md)

	response, err := client.DeployEthContract(ctx, &sidechain_eth.Request{Input: jwtTokenString})
	if err != nil {
		log.Fatalf("Failed to execute 'DeployEthContract' method: %v", err)
	}

	if response.Status == true {
		recvToken, err := jwt.Parse(response.Output, func(recvToken *jwt.Token) (interface{}, error) {
			if _, ok := recvToken.Method.(*jwt.SigningMethodHMAC); !ok {
				return nil, fmt.Errorf("unexpected signing method: %v", recvToken.Header["alg"])
			}
			return []byte(apiKey), nil
		})

		if recvClaims, ok := recvToken.Claims.(jwt.MapClaims); ok && recvToken.Valid {
			strMap := recvClaims["jwt_info"].(map[string]interface{})
			result, _ := json.Marshal(strMap["result"].(map[string]interface{}))
			outputData = string(result)
		} else {
			log.Fatalf("Failed to execute 'DeployEthContract' method: %v", err)
		}
	}
	responseData := Response{outputData, response.Status, response.StatusMessage}
	return responseData
}

func (e *SidechainEth) WatchEthContract(apiKey, did, network, contractAddress, contractName, contractCodeHash string) Response {
	var outputData string
	client := sidechain_eth.NewSidechainEthClient(e.Connection)

	jwtInfo, _ := json.Marshal(JWTInfoSidechainEthWatch{
		Network:          network,
		ContractAddress:  contractAddress,
		ContractName:     contractName,
		ContractCodeHash: contractCodeHash,
	})

	claims := JWTClaim{
		JwtInfo: string(jwtInfo),
		StandardClaims: jwt.StandardClaims{
			ExpiresAt: time.Now().UTC().Add(tokenExpiration).Unix(),
		},
	}

	jwtToken := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	jwtTokenString, err := jwtToken.SignedString([]byte(apiKey))
	if err != nil {
		log.Fatalf("Failed to execute 'WatchEthContract' method: %v", err)
	}
	md := metadata.Pairs("did", did)
	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
	defer cancel()
	ctx = metadata.NewOutgoingContext(ctx, md)

	response, err := client.WatchEthContract(ctx, &sidechain_eth.Request{Input: jwtTokenString})
	if err != nil {
		log.Fatalf("Failed to execute 'WatchEthContract' method: %v", err)
	}

	if response.Status == true {
		recvToken, err := jwt.Parse(response.Output, func(recvToken *jwt.Token) (interface{}, error) {
			if _, ok := recvToken.Method.(*jwt.SigningMethodHMAC); !ok {
				return nil, fmt.Errorf("unexpected signing method: %v", recvToken.Header["alg"])
			}
			return []byte(apiKey), nil
		})

		if recvClaims, ok := recvToken.Claims.(jwt.MapClaims); ok && recvToken.Valid {
			strMap := recvClaims["jwt_info"].(map[string]interface{})
			result, _ := json.Marshal(strMap["result"].(map[string]interface{}))
			outputData = string(result)
		} else {
			log.Fatalf("Failed to execute 'WatchEthContract' method: %v", err)
		}
	}
	responseData := Response{outputData, response.Status, response.StatusMessage}
	return responseData
}
